cli/command/container/run.go

```diff
@@ -159,6 +154,9 @@ func runContainer(ctx context.Context, dockerCli command.Cli, runOpts *runOption
 		defer signal.StopCatch(sigc)
 	}
 
+	ctx, cancelFun := context.WithCancel(context.WithoutCancel(ctx))
+	defer cancelFun()
+
```

Comment:
> 👇 if we do want statusCtx to be a "child" of the context used for attach, it should probably be something like this to not shadow the original context;

Yeah, I believe this is the reason why we have `context.WithCancel(context.WithoutCancel(xx))`

Your suggestion looks good!

Reply:
The suggested change looks alright, although there are some reasons why I don't want to do anymore changes in this PR. The code has become quite difficult to follow (in-terms of which context is used where) since we've been patching out bugs introduced by the context changes. This makes me a bit afraid of doing too many code changes since I don't want to accidentally re-introduce an edge case bug.

By moving line 144 https://github.com/docker/cli/blob/ba1a15433be5fe91edacd09b0c133f5f137ba5e2/cli/command/container/run.go#L144-L145 

down to the `attachContainer` call on line 184 https://github.com/docker/cli/blob/ba1a15433be5fe91edacd09b0c133f5f137ba5e2/cli/command/container/run.go#L181-L184 

we are making it harder to read the code since line 209 uses the `cancelFun` in some cases https://github.com/docker/cli/blob/ba1a15433be5fe91edacd09b0c133f5f137ba5e2/cli/command/container/run.go#L204-L211

Line 204 uses the context (without original cancel) from line 144 https://github.com/docker/cli/blob/ba1a15433be5fe91edacd09b0c133f5f137ba5e2/cli/command/container/run.go#L204

We cancelled the top-level context originally in this commit https://github.com/docker/cli/commit/991b1303daec833d662ec54002d1b34c8805afab to revert the `runContainer` function to a pre-context behavior since there might exist too many unknown edge case behaviors. For that reason I tried to keep as close to the pre-context behavior as possible, by only moving the code stripping the context a few lines down.

Due to the shared `ctx` between line 144 and 184 and 204 and the `cancelFun` used on line 209, it makes me worried that by moving the `ctx` code from line 144 to line 184 (above `attachContainer`) we might have some case where the `cancelFun` is used but is now nil, or where a future change would be more difficult due to more spaghetti usage of `ctx` and `cancelFun` from line 144. 

If it is alright with you @thaJeztah, I would rather we have future PRs that focus on making some of the code more context aware and slowly split up the `runContainer` function so we have less `context` spaghetti.

Reply:
@Benehiko yeah, that's fine; we can move this in a follow-up. It was mostly because the comment described something that happened elsewhere, so I thought moving the code there could help understanding, but it's not critical for sure.